### PR TITLE
Rough draft of a few examples

### PR DESCRIPTION
**Pull request by handrews**
This is just to get the conversation started about this approach
to examples.

Note that the diff is relative to the rewrite.

### CODE REVIEW

File: jsonschema-hyperschema.xml [hunk 1]
```diff
@@ -1137,6 +1137,209 @@ for varname in templateData:
...
+                    The linked hyper-schema defines the API's base URI and provides
+                    two links:  an "about" link to API documentation, and a "self"
+                    link indicating that this is a schema for the base URI.  In this
+                    case the base URI is also the entry point URI.
+                </t>
+                <figure>
+                    <artwork>
+<![CDATA[
+{
+    "$id": "https://schemaexample.com/entry",
```
Comment (dlax) [inline jsonschema-hyperschema.xml]: missing `.`

File: jsonschema-hyperschema.xml [hunk 2]
```diff
@@ -1137,6 +1137,209 @@ for varname in templateData:
...
+            "href": "things/{id}",
+            "hrefRequired": "id",
+            "targetSchema": {"$ref": "#"}
+        }
+    ],
+    "definitions": {
+            "type": "integer",
+            "minimum": 1,
+            "readOnly": true
+        }
```
Comment (dlax) [inline jsonschema-hyperschema.xml]: missing an `id` key in "definitions"

File: jsonschema-hyperschema.xml [hunk 3]
```diff
@@ -1137,6 +1137,209 @@ for varname in templateData:
...
+                    <artwork>
+<![CDATA[{
+    "rel": "tag:rel.example.com,2017:thing",
+    "href": "things/{id}",
+    "hrefSchema": {
+        "required": ["id"],
+        "properties": {
+            "id": {"$ref": "thing#/definitions/id"}
+                "type": "integer",
+                "minimum": 1
```
Comment (dlax) [inline jsonschema-hyperschema.xml]: `"type"` and `"minimum"` should not be there I guess?

File: jsonschema-hyperschema.xml [hunk 4]
```diff
@@ -1137,6 +1137,209 @@ for varname in templateData:
...
+    "required": "data",
+    "properties": {
+        "id": {"$ref": "#/definitions/id"},
+        "data": true
+    },
+    "links": [
+        {
+            "rel": "self",
+            "href": "things/{id}",
+            "hrefRequired": "id",
```
Comment (dlax) [inline jsonschema-hyperschema.xml]: `hrefRequired` must be an array.

File: jsonschema-hyperschema.xml [hunk 5]
```diff
@@ -1137,6 +1137,209 @@ for varname in templateData:
...
+}]]>
+                    </artwork>
+                </figure>
+                <t>
+                    The "href" value here is the same, but everything else is different.
+                    Recall that the instance is an empty object, so "id" cannot be resolved
+                    from instance data.  Instead it is required as client input.  This LDO
+                    could also have used "hrefRequired" but with "required" in "hrefSchema"
+                    it is not strictly necessary.  Providing "hrefRequired" without marking
+                    "id" as requried in "hrefSchema" would lead to errors, as client input
```
Comment (dlax) [inline jsonschema-hyperschema.xml]: typo: requried

File: jsonschema-hyperschema.xml (hunk 3 above)
Comment (handrews) [reply jsonschema-hyperschema.xml]: This is what I get for deciding, fairly late at night, "meh, this is far enough along for feedback" and just pushing it with only a cursory proofread.  Sorry about that :-/

**Comment by philsturgeon** [comment]
Very simple. Start with the most common usages and leaving room to get super technical later. 👍

**Comment by dlax** [comment]
One thing that'd be nice ultimately would be that all examples in the spec share the same "theme". For instance, in OpenAPI, most examples concern a pet store; in JSON API, they're all about articles/author/comments.
I think this helps the reader because they can focus on the specification vocabulary more easily as they get familiar with the underlying theme and they don't have to get through a new theme on each example.

@handrews Just mentioning the idea before you engage into more work on this side.

**Comment by handrews** [comment]
@dlax yeah, that has occurred to me.  Unfortunately that's the part of writing examples that I'm particularly bad at, hence the "thing" schema.  Concrete examples usually feel too contrived to me (plus I find that if there's a concrete example, someone inevitably complains that this isn't how they would have done it, seeming to fail to get the point that an example is just an example and not an endorsement of perfection or whatever).

I really, really hate designing examples.

I think the best way to handle this is for me to work through writing them all the best way that I can manage, which is semi-abstract.  Afterwards maybe someone can come up with a storyline for them, in which case I will be more than happy to rework the examples as needed.

I agree that that is the best way to do it, I just have no clue what will work as my notion of what makes a compelling example has repeatedly been proven to be out-of-step with the vast majority of readers.

**Comment by handrews** [comment]
@dlax @philsturgeon I have added another example about processing multiple kinds of input.  It also covers using an alternate URI scheme (`mailto:`) and media type (`multipart/alternative`, with `text/plain` and `text/html` alternatives modeled as an array, using `contentMediaType` to describe each alternative's format).

I went with this next rather than adding the Collections section as it turns out the the Collections schema basically uses nearly every feature together.  It's probably best done last, although the narrative may be less clear that way.

Again, I'm going to work through this with my own preferred examples, because if I try to come up with a concrete narrative before putting up a PR we're going to be here until 2020.  So I'll go through the rest of the cases and do the collection at the end, and we'll see if we can build a narrative for that once the whole sequence is up.

I'm planning to keep adding commits to this PR rather than stack a whole bunch of PRs, unless someone objects.

**Comment by dlax** [comment]
@handrews 
> I think the best way to handle this is for me to work through writing them all the best way that I can manage, which is semi-abstract. Afterwards maybe someone can come up with a storyline for them, in which case I will be more than happy to rework the examples as needed.

Sure, that seems eminently reasonable!

File: jsonschema-hyperschema.xml [hunk 6]
```diff
@@ -1384,37 +1385,196 @@ Link: <https://schema.example.com/entry> rel=describedBy
...
-                    Forbidding input and requiring instance data.
-                    Requiring input and forbidding instance data.
-                </cref></t>
+            <section title="Submitting a payload and accepting URI input">
+                <t>
+                    This example covers using the "submission" fields for non-representation
+                    input, as well as using them alongside of resolving the URI Template with
+                    input.  Unlike HTML forms, which require either constructing a URI or
+                    sending a payload, but do not allow not both at once, JSON Hyper-Schema can
+                    describe both sorts of input to for the same operation on the same link.
```
Comment (dlax) [inline jsonschema-hyperschema.xml]: typo: "to for"

File: jsonschema-hyperschema.xml [hunk 7]
```diff
@@ -1384,37 +1385,196 @@ Link: <https://schema.example.com/entry> rel=describedBy
...
+                    payload, as seen in the <xref target="HTTP">appendix on HTTP usage</xref>.
+                </t>
+                <t>
+                    In this case, we use a "mailto:" URI, which, per
+                    <xref target="RFC6068">RFC 6068, Section 3"</xref>, does not provide any
+                    operation for retrieving a resource.  It can only be used to construct
+                    a message for sending.  Since there is no concept of a retrievable,
+                    replaceable, or deletable target resource, "targetSchema" and
+                    "targetMediaType" are not used.  Non-representation payloads are
+                    described by "submissionSchema" and "submission MediaType".
```
Comment (dlax) [inline jsonschema-hyperschema.xml]: typo: "submission MediaType".

Also, the last sentence seems superfluous with respect to the next paragraph (and also the first sentence of this section).

File: jsonschema-hyperschema.xml [hunk 8]
```diff
@@ -1384,37 +1385,196 @@ Link: <https://schema.example.com/entry> rel=describedBy
...
+                    "title": {
+                        "type": "string"
+                    },
+                    "cc": {
+                        "type": "string",
+                        "format": "email"
+                    },
+                    "email": false
+                }
+            }
```
Comment (dlax) [inline jsonschema-hyperschema.xml]: missing `,`

File: jsonschema-hyperschema.xml [hunk 9]
```diff
@@ -1384,37 +1385,196 @@ Link: <https://schema.example.com/entry> rel=describedBy
...
+                </figure>
+                <t>
+                    So, given the following instance retrieved from "https://api.example.com/stuff":
+                </t>
+                <figure>
+                    <artwork>
+<![CDATA[{
+    "title": "The Awesome Thing",
+    "stuffWorthEmailingAbout": "Lots of text here...",
+    "email": "someone@exapmle.com"
```
Comment (dlax) [inline jsonschema-hyperschema.xml]: typo: exapmle

**Comment by handrews** [comment]
I'm going to update this with fixes to the last batch of errors @dlax spotted but otherwise this will get merged on Friday to kick off the final review period.  I will open new PRs for further examples.

**Pull request merged**